### ISSUE TITLE
[API Proposal]: `DbDataReaderEnumerator` with `MoveNextAsync`

### ISSUE THREAD

**Issue by wessleym** [dotnet/runtime#98890]
### Background and motivation

I would like to use `DbEnumerator.MoveNextAsync`, which does not exist. (`MoveNext` does exist, but again, I want to read asynchronously.) `DbEnumerator` accepts `IDataReader` as a parameter, and since `IDataReader` does not support `ReadAsync`, `DbEnumerator` cannot be readily given async abilities without breaking changes.
I have made a `DbDataReaderEnumerator` class which requires `DbDataReader` as a parameter. `DbDataReader` has a `ReadAsync` method, so adding `DbDataReaderEnumerator.MoveNextAsync` is trivial. This class just reuses `MoveNext` code to make a `MoveNextAsync` with async enhancements.
This is my suggested file:
https://github.com/wessleym/runtime/blob/d49c266b80dcc36566e55b72d93da2fe1fc98509/src/libraries/System.Data.Common/src/System/Data/Common/DbDataReaderEnumerator.cs

### API Proposal

```csharp
namespace System.Data.Common;

public class DbDataReaderEnumerator : IEnumerator<DbDataRecord>, IAsyncEnumerator<DbDataRecord>
{
    public DbDataReaderEnumerator(DbDataReader reader);
    public DbDataRecord Current { get; }
    public bool MoveNext();
    public async ValueTask<bool> MoveNextAsync();
}
```

### API Usage

```csharp
var readerEnumerator = new DbDataReaderEnumerator(reader);
var anyResults = await readerEnumerator.MoveNextAsync().ConfigureAwait(false);
```

### Alternative Designs

An alternative would be to allow `DbEnumerator` to accept `DbDataReader` as a parameter, but that would introduce breaking changes to callers who were passing an `IDataReader` as a parameter.

### Risks

I am aware of no risks since `DbDataReaderEnumerator` is a new class.

**Comment by roji** [comment]
Some initial thoughts:

* First, adding an enumerator type - as opposed to an IAsyncEnumerable - is less useful; for example, one won't be able to do `await foreach` on it, or use System.Linq.Async to compose LINQ operators on top of it. Is there any particular reason you're proposing adding an enumerator rather than an enumerable?
* While this would allow async moving from record to record (which is a good thing), DbDataRecord itself still has no async methods on it (unlike DbDataReader which does have GetFieldValueAsync and IsDBNullAsync), so this is proposal is incomplete in that sense.
* DbDataReader already implements IEnumerable (with GetEnumerator as abstract), and implementing types typically returning [DbEnumerator](https://learn.microsoft.com/en-us/dotnet/api/system.data.common.dbenumerator?view=net-7.0) (e.g. [SqlClient's GetEnumerator](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs#L1236)). Your proposed DbDataReaderEnumerator seems to be conceptually very similar to this existing DbEnumerator - I'm not sure why it would make sense for a new type to be introduced.

Stepping back, the "enumerability" story of DbDataReader generally isn't great; as I wrote above, DbDataReader implement IEnumerable, but the non-generic IEnumerable only. And once again, the enumerated type - DbDataRecord - hasn't evolved with the rest of ADO.NET and e.g. doesn't support async, as well as probably other APIs. I think that if we want to evolve this enumerability story, we'd need to consider the design more holistically and think of all these problems together, rather than adding a new type such as DbDataReaderEnumerator which addresses only a single concern, etc.

At the same time, this is the first time I'm seeing a complaint about DbDataReader not providing an async enumerator - typical usage is done via DbDataReader.MoveNextAsync() directly. Is there any particular reason you require an actual enumerator here, where MoveNextAsync() isn't sufficient?

**Comment by wessleym** [comment]
1. I agree about an enumerable vs. an enumerator. But `DbEnumerator` allows for the row's data to be saved into the record with a `DataRecordInternal`. When using `DbDataReader.ReadAsync()`, if the reader gets disposed followed by an attempt to read data from the record, an exception will be thrown. So let's say my code reads the records, disposes of the reader, and then tries to read the record values later on. `DbDataReader.ReadAsync()` doesn't allow for that, but `DbEnumerator` does since it caches the data.
2. Since I would prefer that my data get loaded into the record initially like with `DbEnumerator`, I wouldn't need  `GetFieldValueAsync` or `IsDBNullAsync`. I would already have a dictionary of all record objects for fast value retrieval and `DBNull` checks. (Interesting, `SqlDataReader.GetFieldValueAsync` doesn't seem to be an async method anyway, despite the name.)
3. The reason why I would think a new type would be necessary is because `DbEnumerator` requires just an `IDataReader`. `DbDataReaderEnumerator` requires a `DbDataReader`. `DbDataReader` is needed for `ReadAsync()`. I think you would want to allow `DbEnumerator` to continue to accept just an `IDataReader`, but an `IDataReader` is not sufficient for my purposes in this new class.

Regarding your second-to-last paragraph, that sounds great, but it also sounds like a lot of work, and I'm not sure when that would be done. My suggestion just addresses my specific need.

Regarding your last paragraph, please see my point 1 above in this comment.

Thank you for your time and consideration.

**Comment by roji** [comment]
> DbDataReader.ReadAsync() doesn't allow for that, but DbEnumerator does since it caches the data.

I think there's some confusion around this; DbEnumerator doesn't do any caching whatsoever ([source code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Data.Common/src/System/Data/Common/DbEnumerator.cs)) - it's merely an implementation of the IEnumerator interface over DbDataReader. That is, it internally saves a DbDataReader, and calls MoveNext() on it as you enumerate it. Neither does your proposed DbDataReaderEnumerator in #98891.

So in both types, if you dispose the DbDataReader before fully enumerating, you would get an ObjectDisposeException just as if you had accessed DbEnumerator directly. The point of the IEnumerable (and IEnumerator) APIs is not to cache the entire data (which would take up lots of memory), but to the contrary, to enumerate (without any caching). In fact, if DbEnumerator/DbDataReaderEnumerator cached the entire resultset as you seem to be expecting, there would no longer be a need for async methods to enumerate it, since all data would be in memory anyway.

If you're simply looking to cache the entire resultset so that you can dispose DbDataReader early, there are various ways to do that: the easiest would probably to do a small loop over the rows with ReadAsync(), and read each one with DbDataReader.GetValues(). Or you can use something like Dapper or EF, which simply materializes the rows into objects for you.

> Interesting, SqlDataReader.GetFieldValueAsync doesn't seem to be an async method anyway, despite the name

That might be a problem in the SqlClient implementation of that method (SqlClient has various issues with its async support) - but of course System.Data is implemented by other drivers of other databases (the Npgsql implementation definitely is fully async).

> The reason why I would think a new type would be necessary is because DbEnumerator requires just an IDataReader.

It may be possible to retro-fit DbDataReader with DbDataReader support, for the async support. But in any case, we need to consider having DbDataReader implement IAsyncEnumerable, with a default implementation that either returns the new enumerator type (or the retro-fitted DbEnumerator), or throws, or possibly returns a sync enumerator - as most System.Data APIs do in the async default implementations. This was my point about having to consider this holistically rather than just adding a new public type to fix your specific, narrow need.

> Regarding your second-to-last paragraph, that sounds great, but it also sounds like a lot of work, and I'm not sure when that would be done. My suggestion just addresses my specific need.

Understood. I've opened #98903, summarizing some of the points made in this issue, and considering the larger DbDataReader enumerability story. I'll go ahead and close this issue for now - but please feel free to continue the conversation here (e.g. with regards to the caching question above). In any case, if we see more user interest we can definitely look at improving things in this area.

**Comment by roji** [comment]
Re caching, have you tried this with more than one row (and replacing First() with e.g. ToList())? DbEnumerator is probably caching the row it's already one (the first row), but once you call MoveNext() on it, it should just call MoveNext() on the underlying reader and you'll get the same exception

**Comment by wessleym** [comment]
Yes.
```
var records1 = GetRecords("Some Connection String", [], "select UserName from [User]").ToArray();
var records2 = await GetRecordsAsync("Some Connection String", [], "select UserName from [User]", CancellationToken.None).ToArrayAsync().ConfigureAwait(false);
_ = records1[anyIndex]["UserName"];//works
_ = records2[anyIndex]["UserName"];//fails with same InvalidOperationException: "Invalid attempt to call MetaData when reader is closed."
```
I contend the synchronous example works because `DbEnumerator` is saving each record's data into a `DataRecordInternal`:
https://github.com/dotnet/runtime/blob/66aebedbae0bda653933fbb65876770567f04441/src/libraries/System.Data.Common/src/System/Data/Common/DbEnumerator.cs#L68

**Comment by wessleym** [comment]
I agree with your first paragraph and know of nothing I've said in disagreement with it. `DbEnumerator` saves each row's data into a `DataRecordInternal`, thus allowing my `GetRecords` method to work while my `GetRecordsAsync` method fails.
"I'm not sure exactly what's going on in your test code, ...": If I can explain something specific further, I would be happy to do so. But the reader is being disposed before the data gets saved in the case of `GetRecordsAsync`.
As for your second paragraph, I agree I could do it myself, which is why I wrote my `DbDataRecordEnumerator` class, which largely imitates `DbEnumerator`. Outside of a new `DbDataRecordEnumerator` class, I can call `Read()` and `GetValues()` (and then put that data into a performant dictionary of some kind). But I don't need to do that with `DbEnumerator` since .NET has already written a nice implementation for us. I would hope for the same thing in the asynchronous world. And maybe we'll get it some day, but from what I'm reading, it seems I will need to wait for the full approach of #98903.

**Comment by roji** [comment]
No problem @wessleym, I think we had a miscommunication around row caching (what you meant) and resultset caching (what I understood). My only point was that DbEnumerator is just as prone to produce ObjectDisposedException **when enumerating** - it doesn't anything there. Yes, it does cache the row and therefore allows access to the DbDataRecord even if the wrapped DbDataReader has been disposed, but that seems of limited value if one cannot enumerate to the next row.

In any case, this is indeed a case of "nice to have" sugar - there's nothing blocking that this would unblock (in fact, you can have your async enumerator in your own code just as well as in System.Data). I do agree that ideally DbDataReader would have a more complete enumeration story that includes async, but since that's mainly sugar it seems rather low priority (and in any case, must be done with the entire API in mind as per #98903).

**Comment by wessleym** [comment]
Regarding your first paragraph, I think we now both agree that that is not what I was talking about.
Regarding your second paragraph, I'm assuming when you say this is sugar, you're saying this is syntactic sugar. I don't see how this is _just_ syntactic sugar since it allows access to `DataRecordInternal` in an asynchronous way. I guess if you consider `DbEnumerator` to be sugar, then this would be as well. I actually consider `DbEnumerator` to be quite valuable.
"in fact, you can have your async enumerator in your own code just as well as in System.Data": I can't readily have this enumerator in my own code since many of the members of `DbEnumerator` and `DbDataReaderEnumerator` utilize internal classes (e.g., `DataRecordInternal`). My `DbDataReaderEnumerator` could not operate within my own library without replacing the types it uses as well. I can write my own enumerator, but I would want to use the most high-performance algorithms, and I would assume those would be built into .NET.
Still, I understand that this will not be added to .NET. I think that's unfortunate since a ready-to-go record retrieval system (utilizing something akin to `DataRecordInternal`) would be valuable to me, but I guess everyone is just writing their own record dictionary systems. I have done the same, but I would prefer if .NET provided a high-performance system. But maybe that will come in #98903.
Thank you for your time.

**Comment by roji** [comment]
> I would want to use the most high-performance algorithms [...] I would prefer if .NET provided a high-performance system

You talk multiple times about high performance, but I can't see anything like that happening here; DbEnumerator (and DbDataReaderEnumerator) wouldn't be particularly high-performance - they're simply an enumerating interface over DbDataReader. Am I missing something?

Again, I recognize the **convenience** that an IEnumerable/IAsyncEnumerable interface to DbDataReader provide - and I do think this is something we should work on at some point (hence #98903). But I can't see why performance is playing a role here.

**Comment by wessleym** [comment]
If a particular feature is provided by .NET, I tend to assume it has good performance, is being maintained by the .NET community, and over time will be given the latest improvements. In the last few years, I've seen lots of blog posts about great strides in performance for things like file access. And so, if I utilize a .NET feature, my hope that is that I am "subscribed" to decent (and possibly increasing) performance that will generally outperform anything I make myself. Obviously that won't always be true, as in this case. But if I were using `DbEnumerator`, I would hope I would have the fastest possible way to enumerate the results. The more I build my own database enumerator (or any other low-level thing) myself, the more I wonder if I'm missing out on the latest performance improvements .NET's built-in types might provide.
Similarly, when I use System.IO.File, I assume I'm getting better performance than what I could likely make myself since the .NET community put time into making, refining, and benchmarking their code. Perhaps my opinion of .NET is too high, but I really don't think so.
So, for the time being, to reach the best possible performance, I'll use my own system, and I'll look forward to future possibilities. Thank you.
